### PR TITLE
BoxStore

### PR DESCRIPTION
**Pull request by matthewgraf**

### CODE REVIEW

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    pathParts match {
+      case Nil => parentFolder
+      case head :: tail =>
+        val matchingItem = parentFolder.getChildren.asScala
+          .find(_.getName.equalsIgnoreCase(head))
+          .getOrElse(parentFolder.createFolder(head))
+
+        if (!matchingItem.getResource.isInstanceOf[BoxFolder]) {
+          throw new Exception(s"Item '${matchingItem.getName}' exists along path but was not folder")
+        } else {
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: else shouldn't be necessary here

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+          val folder = matchingItem.getResource.asInstanceOf[BoxFolder]
+          putFolderAtPath(folder, tail)
+        }
+    }
+  }
+
+  def splitPath(path: Path): (List[String], String) = {
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: Add javadoc to this method

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    boxFileAtPath(src)
+      .foreach(file => {
+        val dstPath = splitPath(dst)
+        val folder = putFolderAtPath(rootFolder, dstPath._1)
+        file.copy(folder, dstPath._2)
+      })
+  }
+
+  /**
+    * Remove byte for given path.
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: Remove "byte"? Should this be bytes or file

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  def boxFolderAtPath(path: Path): Option[BoxFolder] = {
+    boxItemAtPath(path).flatMap {
+      case folder: BoxFolder => Some(folder)
+      case _ => None
+    }
+  }
+
+
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: rm space

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            val pathKey = if (path.key.takeRight(1).equals("/")) path.key.dropRight(1) else path.key
+            val childKey = s"$pathKey/${itemInfo.getName}"
+            Path(path.root, childKey, Some(itemInfo.getSize), isDir, Some(itemInfo.getModifiedAt))
+          })
+
+      listOneFile = Stream.eval(F.delay(
+        Path(path.root, path.key, Some(item.getInfo.getSize), isDir = false, Some(item.getInfo.getModifiedAt))
+      ))
+
+      s <- if (item.isInstanceOf[BoxFolder]) listFiles else listOneFile
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: It seems like we could check this condition without having to construct the listOneFile stream if it's met
cc: @rolandomanrique 

But maybe something like

``
item.isInstanceOf[BoxFolder].ifM(listFiles, Stream.eval(F.delay { ... }))
``

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  /**
+    * Remove byte for given path.
+    *
+    * @param path to remove
+    * @return F[Unit]
+    */
+  override def remove(path: Path): F[Unit] = F.delay {
+    boxFileAtPath(path).foreach(_.delete())
```
Comment (teddywilson) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: should we be using handleWithError() here?

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+          val folder = matchingItem.getResource.asInstanceOf[BoxFolder]
+          putFolderAtPath(folder, tail)
+        }
+    }
+  }
+
+  def splitPath(path: Path): (List[String], String) = {
+    val fullPath = path.root :: path.key.split("/").toList
+    val pathToParentFolder = fullPath.dropRight(1)
+    val key = fullPath.takeRight(1).head
```
Comment (rolandomanrique) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: Use `fullPath.splitAt(path.size - 1)` instead `dropRight` and `takeRight`

File: box/src/main/scala/blobstore/box/BoxStore.scala (hunk 5 above)
Comment (matthewgraf) [reply box/src/main/scala/blobstore/box/BoxStore.scala]: I can't imagine it takes a significant amount of time to construct that stream. Btw we are doing the same thing in `FileStore.list`

Comment (rolandomanrique) [reply box/src/main/scala/blobstore/box/BoxStore.scala]: You would need a Stream to do `ifM` any ways.

File: box/src/main/scala/blobstore/box/BoxStore.scala (hunk 6 above)
Comment (rolandomanrique) [reply box/src/main/scala/blobstore/box/BoxStore.scala]: we are letting error bubble up to users at this point, we may want to revise error handling across all stores.

File: box/src/main/scala/blobstore/box/BoxStore.scala [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    * List paths. See [[StoreOps.ListOps]] for convenient listAll method.
+    *
+    * @param path to list
+    * @return stream of Paths. Implementing stores must guarantee that returned Paths
+    *         have correct values for size, isDir and lastModified.
+    */
+  override def list(path: Path): fs2.Stream[F, Path] = {
+    for {
+      itemOpt <- Stream.eval(F.delay(boxItemAtPath(path)))
+      item <- if (itemOpt.isDefined) Stream.emit(itemOpt.get) else Stream.empty
```
Comment (stew) [inline box/src/main/scala/blobstore/box/BoxStore.scala]: I'd prefer we never call Option#get, and I made a PR with one suggested alternative. another alternative 
`itemOpt.fold[Stream[F,BoxItem]](Stream.empty)(Stream.emit(_))`

**Comment by matthewgraf** [comment]
I pulled the changes down and ran BoxStoreIntegrationTest ✅

**Pull request merged**